data-prepper-plugins/saas-source-plugins/saas-crawler/src/main/java/org/opensearch/dataprepper/plugins/source/saas_crawler/base/SaasSourcePlugin.java

```diff
@@ -0,0 +1,120 @@
+import org.opensearch.dataprepper.plugins.source.saas_crawler.coordination.partition.LeaderPartition;
+import org.opensearch.dataprepper.plugins.source.saas_crawler.coordination.scheduler.LeaderScheduler;
+import org.opensearch.dataprepper.plugins.source.saas_crawler.coordination.scheduler.WorkerScheduler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Objects;
+import java.util.concurrent.ExecutorService;
+import java.util.function.Function;
+
+
+
+/**
+ * JiraConnector connector entry point.
+ */
+
+public abstract class SaasSourcePlugin implements Source<Record<Event>>, UsesEnhancedSourceCoordination {
+
+
+  private static final Logger log = LoggerFactory.getLogger(SaasSourcePlugin.class);
+  private final PluginMetrics pluginMetrics;
+  private final PluginFactory pluginFactory;
+
+  private final AcknowledgementSetManager acknowledgementSetManager;
+
+  private final ExecutorService executorService;
+
+  private EnhancedSourceCoordinator coordinator;
+
+  @Getter
+  private final SaasSourceConfig sourceConfig;
+
+  private Buffer<Record<Event>> buffer;
+  private final Crawler crawler;
+
+
+  @DataPrepperPluginConstructor
+  public SaasSourcePlugin(final PluginMetrics pluginMetrics,
+                          final SaasSourceConfig sourceConfig,
+                          final PluginFactory pluginFactory,
+                          final AcknowledgementSetManager acknowledgementSetManager,
+                          final Crawler crawler,
+                          final SaasPluginExecutorServiceProvider executorServiceProvider) {
+    log.info("Create SaaS Source Connector");
+    this.pluginMetrics = pluginMetrics;
+    this.sourceConfig = sourceConfig;
+    this.pluginFactory = pluginFactory;
+    this.crawler = crawler;
+
+    this.acknowledgementSetManager = acknowledgementSetManager;
+    this.executorService = executorServiceProvider.get();
+  }
+
+
+  @Override
+  public void start(Buffer<Record<Event>> buffer) {
+    Objects.requireNonNull(coordinator);
+    log.info("Starting SaaS Source Plugin... ");
+    this.buffer = buffer;
+
+    boolean isPartitionCreated = coordinator.createPartition(new LeaderPartition());
+    log.debug("Leader partition creation status: {}", isPartitionCreated);
+
+    Runnable leaderScheduler = new LeaderScheduler(coordinator, this, crawler);
+    this.executorService.submit(leaderScheduler);
+    //Register worker threaders
+    for(int i=0; i<sourceConfig.DEFAULT_NUMBER_OF_WORKERS; i++) {
+      WorkerScheduler workerScheduler = new WorkerScheduler(buffer, coordinator, sourceConfig, crawler);
+      this.executorService.submit(new Thread(workerScheduler));
+    }
+  }
+
+
+  @Override
+  public void stop() {
+    log.info("Stop Source Connector");
+    this.executorService.shutdownNow();
+  }
+
+  @Override
+  public boolean areAcknowledgementsEnabled() {
+    return Source.super.areAcknowledgementsEnabled();
```

Comment:
Not a good idea. This should be left to the derived classes.

Reply:
Agree. Removed this method. Each source plugin will implement their own version.